app/src/main/java/com/google/android/soundchecker/harmonicanalyzer/HarmonicAnalyzer.kt

```diff
@@ -122,6 +118,16 @@ class HarmonicAnalyzer {
         // Calculate Signal To Noise Ratio in dB
         val signalNoisePowerRatio = signalMagSquared / noiseMagSquared
         result.signalNoiseRatioDB = powerToDecibels(signalNoisePowerRatio.toDouble())
+
+        var peakMagSquared = VERY_SMALL_NUMBER
+        for (i in 0 until numFrames / 2) {
+            peakMagSquared = max(peakMagSquared, magnitudeSquared(buffer[i], mImaginary!![i]))
+        }
+        result.bins = FloatArray(numFrames / 2)
+        for (i in 0 until numFrames / 2) {
+            result.bins!![i] = sqrt(magnitudeSquared(buffer[i],
+                    mImaginary!![i]) / peakMagSquared)
```

Comment:
This code calculates the magnitudeSquared twice.
Then it divides a magnitude by a max magnitude squared, which is mathematically invalid.
You could:

    calculate magnitudeSquaredArray into result.bins and find peakMagSquared in one loop
    scaler = 1.0 / sqrt(peakMagSquared)
    normalize by multiplying all of result.bins by scaler

Multiplication is faster than division so best to do the divide outside the loop.

Reply:
> Then it divides a magnitude by a max magnitude squared, which is mathematically invalid.
I think this is fine. sqrt(9/4) = sqrt(9)/sqrt(4) = 3/2

I changed this to a two pass algorithm. Let me know what you think. Thanks!

Reply:
> I think this is fine. sqrt(9/4) = sqrt(9)/sqrt(4) = 3/2

Yes, but the old calculation used a magnitude divided by a magnitudeSquared. So the units don't cancel.

I think the new code is fine.